browser/ui/views/commands/accelerator_service.cc

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/ui/views/commands/accelerator_service.h"
+#include <string>
+#include <utility>
+#include <vector>
+#include "base/containers/contains.h"
+#include "base/containers/cxx20_erase_vector.h"
+#include "base/containers/flat_map.h"
+#include "brave/app/command_utils.h"
+#include "brave/components/commands/common/accelerator_parsing.h"
+#include "brave/components/commands/common/accelerator_pref_manager.h"
+#include "brave/components/commands/common/commands.mojom-forward.h"
+#include "brave/components/commands/common/commands.mojom.h"
+#include "components/prefs/pref_service.h"
+#include "mojo/public/cpp/bindings/pending_remote.h"
+#include "ui/base/accelerators/accelerator.h"
+
+namespace commands {
+
+namespace {
+
+mojom::CommandPtr ToMojoCommand(
+    int command_id,
+    const std::vector<ui::Accelerator> accelerators) {
+  auto command = mojom::Command::New();
+  command->id = command_id;
+  command->name = commands::GetCommandName(command_id);
+  for (const auto& accelerator : accelerators) {
+    auto a = mojom::Accelerator::New();
+    a->codes = commands::ToCodesString(accelerator);
+    a->keys = commands::ToKeysString(accelerator);
+    command->accelerators.push_back(std::move(a));
+  }
+  return command;
+}
+
+base::flat_map<int, mojom::CommandPtr> ToMojoCommands(
+    base::flat_map<int, std::vector<ui::Accelerator>> commands) {
+  base::flat_map<int, mojom::CommandPtr> result;
+  for (const auto& [command_id, accelerators] : commands) {
+    result[command_id] = ToMojoCommand(command_id, accelerators);
+  }
+  return result;
+}
+}  // namespace
+
+AcceleratorService::AcceleratorService(PrefService* pref_service,
+                                       const Accelerators& default_accelerators)
+    : pref_manager_(pref_service), default_accelerators_(default_accelerators) {
+  Initialize();
+}
+
+AcceleratorService::~AcceleratorService() = default;
+
+void AcceleratorService::Initialize() {
+  auto accelerators = pref_manager_.GetAccelerators();
+  if (accelerators.size() == 0) {
```

Comment:
This looks good enough for now. But I'm wondering what would happen if GetAcceleratorList() returns a different list one day. Something could be gone or added. Should we compare command_ids one by one and add or remove it from pref_manager?

Reply:
Hmm, that's a good point. The problem is that users could also have removed an existing command. I guess we'll need to make a copy of the |DefaultCommands| in the prefs and compare what's changed (if anything) at startup.

As this PR is already pretty big, I've added a followup issue:
https://github.com/brave/brave-browser/issues/28766